Wrong results when filtering on data with null values

### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of polars.


### Issue Description

In certain circumstances, if the columns used in a filter expression contain null values, the query returns zero rows when a non-zero number of rows should be returned.

The issue occurs on the most recent commit on `master`, 7cf9960df019a4976ba027a2848347ec46b1be26 as well as on the last stable release at tag `py-polars-v0.14.9`.

This issue affects both Rust and Python.


### Reproducible Example

```rust
// Tested on polars at commit
// `7cf9960df019a4976ba027a2848347ec46b1be26`.
#[test]
fn test_filter_null() {
  use polars::prelude::*;
  let df = df![
    "x" => [Some(0), Some(1), None, Some(2)],
    "y" => [Some(1), Some(2), None, Some(3)],
  ]
  .unwrap()
  .lazy()
  .filter(col("x").neq(col("y")))
  .collect()
  .unwrap();
  assert_eq!(df.shape(), (3, 2));
}
```

```
---- test_filter_null stdout ----
thread 'test_filter_null' panicked at 'assertion failed: `(left == right)`
  left: `(0, 2)`,
 right: `(3, 2)`', src/lib.rs:32:3
```

```python
import polars as pl
import unittest

# Tested on py-polars at commit
# `7cf9960df019a4976ba027a2848347ec46b1be26` and at tag
# `py-polars-v0.14.9`.
class TestBug(unittest.TestCase):
  def test_bug(self):
    df = pl.DataFrame({
      "x": [0, 1, None, 2],
      "y": [1, 2, None, 3],
    }).filter(pl.col("x") != pl.col("y"))
    self.assertEqual(df.shape, (3, 2))

unittest.main(exit=False)
```

```
AssertionError: Tuples differ: (0, 2) != (3, 2)
```


### Expected Behavior

N/A

### Installed Versions

```toml
features = ["lazy"]
```

--- a/polars/polars-arrow/Cargo.toml
+++ b/polars/polars-arrow/Cargo.toml
@@ -9,9 +9,9 @@ description = "Arrow interfaces for Polars DataFrame library"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd", features = ["compute_concatenate"], default-features = false }
+# arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd", features = ["compute_concatenate"], default-features = false }
 # arrow = { package = "arrow2", path = "../../../arrow2", features = ["compute_concatenate"], default-features = false }
-# arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "comparison_and_validity", features = ["compute_concatenate"], default-features = false }
+arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "comparison_and_validity", features = ["compute_concatenate"], default-features = false }
 # arrow = { package = "arrow2", version = "0.12", default-features = false, features = ["compute_concatenate"] }
 hashbrown = "0.12"
 num = "^0.4"

--- a/polars/polars-core/Cargo.toml
+++ b/polars/polars-core/Cargo.toml
@@ -173,11 +173,11 @@ thiserror = "^1.0"
 
 [dependencies.arrow]
 package = "arrow2"
-git = "https://github.com/jorgecarleitao/arrow2"
-# git = "https://github.com/ritchie46/arrow2"
-rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd"
+# git = "https://github.com/jorgecarleitao/arrow2"
+git = "https://github.com/ritchie46/arrow2"
+# rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd"
 # path = "../../../arrow2"
-# branch = "comparison_and_validity"
+branch = "comparison_and_validity"
 # version = "0.12"
 default-features = false
 features = [

--- a/polars/polars-io/Cargo.toml
+++ b/polars/polars-io/Cargo.toml
@@ -38,8 +38,8 @@ private = ["polars-time/private"]
 [dependencies]
 ahash = "0.7"
 anyhow = "1.0"
-arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd", default-features = false }
-# arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "comparison_and_validity", default-features = false }
+# arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "6c102a0c3e2dbeb185360dd3d5c3637b5e2028fd", default-features = false }
+arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "comparison_and_validity", default-features = false }
 # arrow = { package = "arrow2", version = "0.12", default-features = false }
 # arrow = { package = "arrow2", path = "../../../arrow2", default-features = false }
 csv-core = { version = "0.1.10", optional = true }

--- a/py-polars/tests/unit/test_queries.py
+++ b/py-polars/tests/unit/test_queries.py
@@ -377,3 +377,13 @@ def test_query_4538() -> None:
     assert df.select([pl.col("value").str.to_uppercase().is_in(["AAA"])])[
         "value"
     ].to_list() == [True, False]
+
+
+def test_none_comparison_4773() -> None:
+    df = pl.DataFrame(
+        {
+            "x": [0, 1, None, 2],
+            "y": [1, 2, None, 3],
+        }
+    ).filter(pl.col("x") != pl.col("y"))
+    assert df.shape == (3, 2)

